src/ol_infrastructure/components/services/k8s.py

```diff
@@ -350,3 +362,365 @@ def __init__(
+    ):
+        super().__init__(
+            "ol:infrastructure:services:k8s:OLApisixSharedPlugin", name, None, opts
+        )
+
+        __default_plugins: list[dict[str, Any]] = [
+            {
+                "name": "redirect",
+                "enable": True,
+                "config": {
+                    "http_to_https": True,
+                },
+            },
+            {
+                "name": "cors",
+                "enable": True,
+                "config": {
+                    "allow_origins": "**",
+                    "allow_methods": "**",
+                    "allow_headers": "**",
+                    "allow_credential": True,
+                },
+            },
+            {
+                "name": "response-rewrite",
+                "enable": True,
+                "config": {
+                    "headers": {
+                        "set": {
+                            "Referrer-Policy": "origin",
+                        }
+                    },
+                },
+            },
+        ]
+
+        resource_options = pulumi.ResourceOptions(parent=self).merge(opts)
+
+        if plugin_config.enable_defaults:
+            plugin_config.plugins.extend(__default_plugins)
+
+        self.resource_name = (
+            f"{plugin_config.application_name}-{plugin_config.resource_suffix}"
+        )
+        self.shared_plugin_apisix_pluginconfig_resource = (
+            kubernetes.apiextensions.CustomResource(
+                f"OLApisixSharedPlugin-{self.resource_name}",
+                api_version="apisix.apache.org/v2",
+                kind="ApisixPluginConfig",
+                metadata=kubernetes.meta.v1.ObjectMetaArgs(
+                    name=self.resource_name,
+                    labels=plugin_config.k8s_labels,
+                    namespace=plugin_config.k8s_namespace,
+                ),
+                spec={
+                    "plugins": plugin_config.plugins,
+                },
+                opts=resource_options,
+            )
+        )
+
+
+class OLApisixExternalUpstreamConfig(BaseModel):
+    application_name: str
+    resource_suffix: str = "external-upstream"
+    k8s_labels: dict[str, str] = {}
+    k8s_namespace: str
+    external_hostname: str
+    scheme: str = "https"
+
+
+class OLApisixExternalUpstream(pulumi.ComponentResource):
+    """
+    External upstream configuration for apisix
+    Defines and creates an "ApisixUpstream" resource in the k8s cluster
+    This is for a service that is hosted outside of kubernetes but we want
+    to have APISIX in front of it anyways.
+    """
+
+    def __init__(
+        self,
+        name: str,
+        external_upstream_config: OLApisixExternalUpstreamConfig,
+        opts: Optional[pulumi.ResourceOptions] = None,
+    ):
+        super().__init__(
+            "ol:infrastructure:services:k8s:OLApisixExternalUpstream", name, None, opts
+        )
+        resource_options = pulumi.ResourceOptions(parent=self).merge(opts)
```

Comment:
I think that this merge pattern may actually be counterproductive because it causes some uncertainty in terms of resource parenting for nested resources. I removed that in the DB and Vault mount components because of that.

Reply:
Okay! Will update

Reply:
Actually I will do it another time, I am doing a TLS / Certmanager resource next and will get it then.